charger/pulsatrix.go

```diff
@@ -0,0 +1,269 @@
+}
+
+// ConnectWs connects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) connectWs() error {
+	ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
+	defer cancel()
+	c.ctx = ctx
+	c.log.INFO.Printf("connecting to %s", c.uri)
+	conn, _, err := websocket.Dial(ctx, c.uri, nil)
+	if err != nil {
+		if strErr := fmt.Sprintf("%s", err); strings.Contains(strErr, "context deadline exceeded") {
+			err = fmt.Errorf("Make sure the IP is correct and the SECC is connected to the network")
+		} else {
+			err = fmt.Errorf("error connecting to websocket: %v", err)
+		}
+		return err
+	}
+	c.updated = time.Now()
+	c.enState = false
+	c.conn = conn
+	c.handleError(c.Enable(false))
+	go c.wsReader()
+	go c.heartbeat()
+	return nil
+}
+
+// ReconnectWs reconnects to a pulsatrix SECC websocket
+func (c *PulsatrixCharger) reconnectWs() {
+	c.handleError(backoff.RetryNotify(func() error {
+		err := c.connectWs()
+		if err != nil {
+			return err
+		}
+	}, c.bo, func(err error, duration time.Duration) {
+		c.log.WARN.Printf("trying to reconnect in %v...\n", duration)
+	}))
+}
+
+// WsReader runs a loop that reads messages from the websocket
+func (c *PulsatrixCharger) wsReader() {
+	go func() {
+		for c.valid() {
+			ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
+			defer cancel()
+			messageType, message, err := c.conn.Read(ctx)
+			if err != nil {
+				c.log.ERROR.Println("error reading message:", err)
+				break
+			} else {
+				c.parseWsMessage(messageType, message)
+				c.updated = time.Now()
+			}
+		}
+		c.mutex.Lock()
+		if c.conn != nil {
+			c.conn.Close(websocket.StatusNormalClosure, "Reconnecting")
+			c.conn = nil
+		}
+		c.mutex.Unlock()
+		c.reconnectWs()
+	}()
+}
+
+// wsWrite writes a message to the websocket
+func (c *PulsatrixCharger) wsWriter(message string) error {
+	if c.valid() && c.conn != nil {
+		c.mutex.Lock()
+		defer c.mutex.Unlock()
+		ctx, cancel := context.WithTimeout(context.Background(), request.Timeout)
+		defer cancel()
+		err := c.conn.Write(ctx, websocket.MessageText, []byte(message))
+		if err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("Waiting for reconnect")
+	}
+	return nil
+}
+
+// ParseWsMessage parses a message from the websocket
+func (c *PulsatrixCharger) parseWsMessage(messageType websocket.MessageType, message []byte) {
+	if messageType == websocket.MessageText {
+		b := bytes.ReplaceAll(message, []byte(":NaN"), []byte(":null"))
+		stadtIdx := bytes.IndexByte(b, '{')
+		c.handleError(json.Unmarshal(b[stadtIdx:], c))
+	}
+}
+
+// Heartbeat sends a heartbeat to the pulsatrix SECC
+func (c *PulsatrixCharger) heartbeat() {
+	for range time.Tick(3 * time.Minute) {
+		if c.enState {
+			c.handleError(c.Enable(true))
+		} else {
+			c.handleError(c.Enable(false))
+		}
+	}
+}
+
+func (c *PulsatrixCharger) valid() bool {
+	return !c.updated.Before(time.Now().Add(-30 * time.Second))
+}
+
+func (c *PulsatrixCharger) handleError(err error) {
```

Comment:
do we need this as function?

Reply:
I thought this would allow for a cleaner code

Reply:
most case you log and return error. logging _is_ error handling, same as returning. you want to handle it only once. If you remove the logging there won't be much uses for this function?